[ENH] Add ability to strip output via regex

peytondmurray opened the pull request:
This PR adds a `widget_regex` parameter to `stripout` to be matched against cells with widget output. Cells with `text/plain` output that matches the regex will be stripped. If no argument is specified for `widget_regex`, or `widget_regex=None`, "^Output();?$" will be used as the default regex.

Note that the `pyo3` version was bumped to `0.25.0` to make use of of `#[pyo3(signature = ...)]` for setting default value of `widget_regex`.

Closes #20.

Blocked by #23.

peytondmurray commented:
@mlucool Does this approach look like it implements what you need to strip arbitrary widget output?

mlucool commented:
[reproducer.json](https://github.com/user-attachments/files/20619981/reproducer.json)
Attaching a notebook with an example (rename to .ipynb). In this example, all of the widgets.Output() should be removed but nothing else when `--keep-output` and `--widget-regex` is used.

So two issues are:
1. There can be multiple outputs to cell, we only want to remove some of them
2. I don't think --widget-regex should be a default. Also feels like it should be something more generic as its not a widget regexp, but something in a execution result?

peytondmurray commented:
Still working on this, but one concern is that in your reproducer notebook, none of the outputs that you might want to strip are of `"output_type": "execute_result"` - they're all either "stream" or "display_data". So maybe it just makes sense to match any output against the regex?

An alternative would be for the user to have to specify a mapping, `{output_type: regex_to_match}`.

> I don't think --widget-regex should be a default. Also feels like it should be something more generic as its not a widget regexp, but something in a execution result?

Maybe? We're specifically targeting cell outputs with a `data["application/vnd.jupyter.widget-view+json"]` field that match the regex. So in that sense we are specifically targeting widgets. We can of course _only_ look at the outputs and target anything matches the regex if that's preferred?

> In this example, all of the widgets.Output() should be removed but nothing else when --keep-output and --widget-regex is used

From a user's perspective, should `--keep-output` override `--widget-regex` (or whatever we end up calling it)? They seem contradictory, but as long as it's well documented I think it could be fine.

mlucool commented:
> So maybe it just makes sense to match any output against the regex?

Ok

> From a user's perspective, should --keep-output override --widget-regex (or whatever we end up calling it)? They seem contradictory, but as long as it's well documented I think it could be fine.

This should be read as "Keep all output, except for this regex"

peytondmurray commented:
Great, the feedback was helpful. I've modified the logic here to examine each output for each cell individually, and modified it so that the regex takes precedence over the `keep_output` option or `'keep_output'` in cell metadata/tags. Here's pseudocode for each cell:

```
for output in cell['outputs']:
    if regex.matches(output['text']):
        output.remove()
    else if has_keep_output_tag || has_keep_output_metadata:
        output.keep()
    else:
        output.keep() if `keep_output` else output.remove()
```

Self-review complete, so I'm marking this as ready once again.

mlucool commented on src/stripoutlib.rs:
```diff
@@ -54,11 +74,85 @@ fn determine_keep_output(cell: &JSONMap, default: bool) -> Result<bool, String>
+                .ok_or("Could not get cell outputs.")?;
+
+            for output in outputs {
+                let obj = output.as_object().ok_or("Cell output is not a map; notebook is malformed.")?;
```
Nitpick: it's an object in JSON (not a map)

mlucool commented on src/main.rs:
```diff
@@ -78,6 +78,10 @@ struct Cli {
     ignore_git_nb_config: bool,
 
+    #[clap(long, action)]
+    /// Specify a regex to use to strip matching output
```
This makes --help confusing. Can we explain a bit more of the order and precedence and maybe include an example in the README or --help of how one may want to use it (our example is removing closed widgets).

peytondmurray replied:
Good point - here's the updated help text:

> Specify a regex to use to strip matching output. Even if other settings would keep cell output (for example, if `--keep-output` is specified), matching cell outputs will be stripped regardless.

I've also added a few examples to the `README.md`, including one about stripping `^Output()$`, and another about removing only completed `tqdm` progress bars (which we have in some of our test notebooks).

mlucool reviewed (commented):
Tests look right now. As noted, let's improve documentation too

mlucool commented on src/main.rs:
```diff
@@ -78,6 +78,12 @@ struct Cli {
     ignore_git_nb_config: bool,
 
+    #[clap(long, action)]
+    /// Specify a regex to use to strip matching output. Even if other settings would keep
```
nitpick: maybe note it's the rust dialect? Not obvious from how people install/use this

mlucool commented:
You can also update the CHANGELOG and release
